File: packages/core/src/factories/JsonMetadataFactory.ts
import { ValidationPipe } from "@typia/interface";
import ts from "@typescript/native-preview";

import { TransformerError } from "../context/TransformerError";
import { AtomicPredicator } from "../programmers/helpers/AtomicPredicator";
import { MetadataCollection } from "../schemas/metadata/MetadataCollection";
import { MetadataSchema } from "../schemas/metadata/MetadataSchema";
import { MetadataFactory } from "./MetadataFactory";

export namespace JsonMetadataFactory {
  export interface IProps {
    method: string;
    checker: ts.TypeChecker;
    transformer?: ts.TransformationContext;
    type: ts.Type;
    validate?: MetadataFactory.Validator;
  }
  export interface IOutput {
    collection: MetadataCollection;
    metadata: MetadataSchema;
  }

  export const analyze = (props: IProps): IOutput => {
    const collection: MetadataCollection = new MetadataCollection();
    const result: ValidationPipe<MetadataSchema, MetadataFactory.IError> =
      MetadataFactory.analyze({
        checker: props.checker,
        transformer: props.transformer,
        options: {
          absorb: true,
          escape: true,
          constant: true,
          validate: props.validate
            ? (next) => {
                const errors: string[] = validate(next);
                errors.push(...props.validate!(next));
                return errors;
              }
            : (next) => validate(next),
        },
        components: collection,
        type: props.type,
      });
    if (result.success === false)
      throw TransformerError.from({
        code: props.method,
        errors: result.errors,
      });
    return {
      collection,
      metadata: result.data,
    };
  };

  export const validate = (props: {
    metadata: MetadataSchema;
    explore: MetadataFactory.IExplore;
  }): string[] => {
    const output: string[] = [];
    if (
      props.metadata.atomics.some((a) => a.type === "bigint") ||
      props.metadata.constants.some((c) => c.type === "bigint")
    )
      output.push("JSON does not support bigint type.");
    if (
      props.metadata.tuples.some((t) =>
        t.type.elements.some((e) => e.isRequired() === false),
      ) ||
      props.metadata.arrays.some((a) => a.type.value.isRequired() === false)
    )
      output.push("JSON does not support undefined type in array.");
    if (props.metadata.maps.length)
      output.push("JSON does not support Map type.");
    if (props.metadata.sets.length)
      output.push("JSON does not support Set type.");
    for (const native of props.metadata.natives)
      if (
        AtomicPredicator.native(native.name) === false &&
        native.name !== "Date"
      )
        output.push(`JSON does not support ${native.name} type.`);
    return output;
  };
}

File: packages/core/src/factories/JsonMetadataFactory.go
package factories

type jsonMetadataFactory struct{}

var JsonMetadataFactory jsonMetadataFactory

func (jsonMetadataFactory) Analyze(props JsonMetadataAnalyzeProps) (JsonMetadataOutput, error) {
	collection := NewMetadataCollection()
	result := MetadataFactory.Analyze(MetadataAnalyzeProps{
		Checker:    props.Checker,
		Components: collection,
		Type:       props.Type,
		Options: MetadataFactoryOptions{
			Absorb:   boolPtr(true),
			Escape:   boolPtr(true),
			Constant: boolPtr(true),
			Validate: func(next MetadataValidateProps) []string {
				errors := JsonMetadataFactory.Validate(next)
				if props.Validate != nil {
					errors = append(errors, props.Validate(next)...)
				}
				return errors
			},
		},
	})
	if !result.Success {
		return JsonMetadataOutput{}, TransformerErrorFrom(props.Method, result.Errors)
	}
	return JsonMetadataOutput{Collection: collection, Metadata: result.Data}, nil
}

func (jsonMetadataFactory) Validate(props MetadataValidateProps) []string {
	output := make([]string, 0)
	for _, atomic := range props.Metadata.Atomics {
		if atomic.Type == "bigint" {
			output = append(output, "JSON does not support bigint type.")
			break
		}
	}
	for _, constant := range props.Metadata.Constants {
		if constant.Type == "bigint" {
			output = append(output, "JSON does not support bigint type.")
			break
		}
	}
	for _, tuple := range props.Metadata.Tuples {
		for _, elem := range tuple.Type.Elements {
			if !elem.IsRequired() {
				output = append(output, "JSON does not support undefined type in array.")
				break
			}
		}
	}
	for _, array := range props.Metadata.Arrays {
		if !array.Type.Value.IsRequired() {
			output = append(output, "JSON does not support undefined type in array.")
			break
		}
	}
	if len(props.Metadata.Maps) != 0 {
		output = append(output, "JSON does not support Map type.")
	}
	if len(props.Metadata.Sets) != 0 {
		output = append(output, "JSON does not support Set type.")
	}
	for _, native := range props.Metadata.Natives {
		if !atomicNative(native.Name) && native.Name != "Date" {
			output = append(output, "JSON does not support "+native.Name+" type.")
		}
	}
	return output
}
